# Variables - Exercise Solutions

## Instructions

In this workbook, you will practice using variables in several new contexts. Additionally, you will learn a few new concepts that make variables even more useful.

For reference, below is the type casting table from the previous video:

| **Type cast** | **What can be casted** | **Possible errors** |
|---------------|------------------------|---------------------|
| From `float` To `str` | All floats can be successfully casted to strings | No type cast error possible |
| From `str` To `float` | Only strings that are float values can be successfully casted to floats, ex `4.3` | Error on strings that are not float values, ex `hello` |
| From `int` To `str` | All integers can be successfully casted to strings | No type cast error possible |
| From `str` To `int` | Only strings that are integer values can be successfully casted to integers, ex `4` | Error on strings that are not integer values, ex `hello` |
| From `int` To `float` | All integers can be successfully casted to floats | No type cast error possible |
| From `float` To `int` | Only floats with `.0` as their decimal component will not lose information on cast | No error, but information loss when floats do not end in `.0` |


### 🚨 New Concept: Off-limits variable names

As we have seen so far, we can choose a wide variety of names for our variables. However, Python reserves some names for itself. These are called keywords. Words like `if`, `else`, `and`, `or`, `in`, and `for` are not allowed to be used in variable names (a full list of the keywords can be found here: https://www.w3schools.com/python/python_ref_keywords.asp). In the rest of this course we will learn how to use many of these keywords. For now, just know that if you try to use any of these keywords for variable names, Python will give you an error.

In [ ]:
in = "my string"

SyntaxError: invalid syntax (<ipython-input-17-6ef55fb8791d>, line 1)

## Exercise 1 - Strings and numeric data don't mix

You have just landed a job as the official stats keeper for the championship basketball game between the Blaze and the Titans. Because you will have many statistics to track, you will use variables to help you track and organize the game's data.

### Part A

First, let's set up two variables to track the two teams' overall wins. So far, the Blaze has won 13 games and the Titans have won 12 games. Create a variables called `blaze_wins` and `titans_wins` and assign their respective number of wins.

**Solution**

In [ ]:
blaze_wins = 13
titans_wins = 12

### Part B

Create two other variables, `blaze_text` and `titans_text`, with the text `"Blaze wins: "` and `"Titan wins: "` respectively.

Then concatenate the Blaze wins variables and the Titans wins variables such that they produce the following output: `Blaze wins: 13, Titan wins: 12`

**Solution**

This one is a bit tricky! Remember, Python does not allow us to concatenate strings and numeric data directly. We must cast our numeric variables to strings with the `str()` function.

In [ ]:
blaze_score_text = "Blaze wins: "
titans_score_text = "Titan wins: "

blaze_score_text + str(blaze_wins) + ", " + titans_score_text + str(titans_wins)

'Blaze wins: 13, Titan wins: 12'

### Part C

### 🚨 New Concept: F-strings

While writing code like `blaze_score_text + str(blaze_wins) + ", " + titans_score_text + str(titans_wins)` does work, it is not the most streamlined way to embed variable values in strings. To improve upon this, Python provides f-strings (also called formatted string literals). We can create an f-string by simply writing the character `f` before the first quotation marks of the string. Then, we can include variables anywhere within the string by placing them within curly braces `{}`. For example...

In [ ]:
month = "January"
day = 1

"Today is " + month + " " + str(day) + "!"

'Today is January 1!'

Can be written with a single f-string like this...

In [ ]:
f"Today is {month} {day}!"

'Today is January 1!'

Using f-strings makes code more readable than using the concatenation operator `+` for joining strings. Additionally, when using f-strings, we are not required to cast numeric values to strings.

### Question

Rewrite the string `Blaze wins: 13, Titan wins: 12` using a single f-string.

**Solution**

With the score text variables...

In [ ]:
f"{blaze_score_text}{blaze_wins}, {titans_score_text}{titans_wins}"

'Blaze wins: 13, Titan wins: 12'

Or without using the score text variables...

In [ ]:
f"Blaze wins: {blaze_wins}, Titan wins: {titans_wins}"

'Blaze wins: 13, Titan wins: 12'

## Exercise 2 - Multiple assignments

### 🚨 New Concept: Multiple assignments

Python allows us to assign multiple values to multiple variables on the same line of code. This is a convenient way to make your code more concise. For example...

In [ ]:
# Assigns "Red" to x, "Green" to y, and "Blue" to z
x, y, z = "Red", "Green", "Blue"
f"x = {x}, y = {y}, z = {z}"

'x = Red, y = Green, z = Blue'

You can also assign the same value to each variable like this...

In [ ]:
# Assigns purple to a, b, and c
a = b = c = "Purple"
f"a = {a}, b = {b}, c = {c}"

'a = Purple, b = Purple, c = Purple'

### Question

At the beginning of the championship game, both the Blaze and the Titans start with a score of 0. Create two variables `titans_score` and `blaze_score` and assign them both to 0. Use multiple assignment to do this on a single line of code.

**Solution**

In [ ]:
titans_score = blaze_score = 0
f"Titans score: {titans_score}, Blaze score: {blaze_score}"

'Titans score: 0, Blaze score: 0'

## Exercise 3 - Variable mutability and reassignment

### 🚨 New Concept: Variable mutability

Python allows us to update the values stored in our variables during our program. This feature is called "mutability". It looks like this...

In [ ]:
my_var = "my string"
my_var

'my string'

In [ ]:
my_var = 3
my_var

3

Here Python allowed us to reassign an integer value to the string variable `my_var`.

### 🚨 New Concept: Other assignment operators

In addition to the `=` assignment operator, Python provides several other assignment operators which serve as shorthand for lines of code like this `x = x + 2`. These assignment operators include `+=`, `-=`, `*=`, `/=`, and many others (that can be found here: https://www.w3schools.com/python/gloss_python_assignment_operators.asp). For example, line 2 in the following cell can be rewritten using the `+=` assignment operator...

In [ ]:
x = 1
x = x + 2
x

3

like this...

In [ ]:
x = 1
x += 2
x

3

### Question

The championship game is underway! It is your job to keep track of the scores.

1. Points scored in the first quarter: Blaze - 23, Titans - 20
2. Points scored in the second quarter: Blaze - 30, Titans - 28
3. Points scored in the third quarter: Blaze - 29, Titans - 30
4. Points scored in the fourth quarter: Blaze - 26, Titans - 34

Using the variables `blaze_score` and `titans_score` and the `+=` assignment operator, update the scores for each quarter of the game. Then output the final scores of each team.

**Solution**

In [ ]:
# 1. Points scored in the first quarter: Blaze - 23, Titans - 20
blaze_score += 23
titans_score += 20

# 2. Points scored in the second quarter: Blaze - 30, Titans - 28
blaze_score += 30
titans_score += 28

# 3. Points scored in the third quarter: Blaze - 29, Titans - 30
blaze_score += 29
titans_score += 30

# 4. Points scored in the fourth quarter: Blaze - 26, Titans - 34
blaze_score += 26
titans_score += 34

f"Titans score: {titans_score}, Blaze score: {blaze_score}"

'Titans score: 112, Blaze score: 108'

Final score - Titans: 112, Blaze: 108

The Titans win!!

## Exercise 4 - Advanced

During the game, your coworker was also keeping track of the Titan star player's stats. Your coworker has their own custom metric that they call "Player Magic". Given the following Player Magic starting value and code that updates the Player Magic during the game, determine (without running the code) what the Player Magic value is at the end of the game.

**Solution**

In [ ]:
player_magic = 10
player_magic %= 2
player_magic += 23
player_magic /= 5
player_magic //= 1
player_magic *= 9
player_magic -= 34
player_magic **= 3

# What is player_magic?
player_magic

8.0